src/ripple/shamap/SHAMapItem.h

```diff
@@ -139,46 +143,49 @@ intrusive_ptr_add_ref(SHAMapItem const* x)
 inline void
 intrusive_ptr_release(SHAMapItem const* x)
 {
+    assert(x);
+
     if (--x->refcount_ == 0)
     {
         auto p = reinterpret_cast<std::uint8_t const*>(x);
 
-        // The SHAMapItem constuctor isn't trivial (because the destructor
-        // for CountedObject isn't) so we can't avoid calling it here, but
-        // plan for a future where we might not need to.
-        if constexpr (!std::is_trivially_destructible_v<SHAMapItem>)
-            std::destroy_at(x);
+        // We need to invoke the destructor for this object before we
+        // release the memory.
+        std::destroy_at(x);
 
-        // If the slabber doens't claim this pointer, it was allocated
+        // If the slabber doesn't claim this pointer, it was allocated
         // manually, so we free it manually.
         if (!detail::slabber.deallocate(const_cast<std::uint8_t*>(p)))
             delete[] p;
     }
 }
 
-inline boost::intrusive_ptr<SHAMapItem>
+inline boost::intrusive_ptr<SHAMapItem const>
 make_shamapitem(uint256 const& tag, Slice data)
 {
     assert(data.size() <= megabytes<std::size_t>(16));
 
     std::uint8_t* raw = detail::slabber.allocate(data.size());
 
-    // If we can't grab memory from the slab allocators, we fall back to
-    // the standard library and try to grab a precisely-sized memory block:
+    // If we can't grab memory from the slab allocators, we use new to
+    // get a precisely-sized and properly aligned memory block:
     if (raw == nullptr)
-        raw = new std::uint8_t[sizeof(SHAMapItem) + data.size()];
+        raw = new (std::align_val_t(alignof(SHAMapItem)))
+            std::uint8_t[sizeof(SHAMapItem) + data.size()];
+
+    assert(reinterpret_cast<std::uintptr_t>(raw) % alignof(SHAMapItem) == 0);
 
     // We do not increment the reference count here on purpose: the
     // constructor of SHAMapItem explicitly sets it to 1. We use the fact
     // that the refcount can never be zero before incrementing as an
     // invariant.
-    return {new (raw) SHAMapItem{tag, data}, false};
+    return {
+        new (std::assume_aligned<alignof(SHAMapItem)>(raw))
+            SHAMapItem{tag, data},
+        false};
 }
 
-static_assert(alignof(SHAMapItem) != 40);
-static_assert(alignof(SHAMapItem) == 8 || alignof(SHAMapItem) == 4);
```

Comment:
@seelabs  There was an assert pertaining to alignment in a previous commit, but I don't think it helps solve the clang warning.

Reply:
It won't. The problem is that on an LP64 system, pointers are aligned on 8 byte boundaries, and `FreeItem` contains a pointer so its natural alignment is 8 bytes, but `SHAMapItem` doesn't contain any pointers, and its natural alignment is 4 bytes.